containers/src/Data/Map/Internal.hs

```diff
@@ -3535,8 +3536,21 @@ argSet (Bin sz kx x l r) = Set.Bin sz (Arg kx x) (argSet l) (argSet r)
 -- > fromSet undefined Data.Set.empty == empty
 
 fromSet :: (k -> a) -> Set.Set k -> Map k a
-fromSet _ Set.Tip = Tip
-fromSet f (Set.Bin sz x l r) = Bin sz x (f x) (fromSet f l) (fromSet f r)
+fromSet f = runIdentity . fromSetA (pure . f)
```

Comment:
```suggestion
fromSet f = runIdentity . fromSetA (coerce f)
```

We don't want to build a new closure that indirectly calls `f`, which is what `pure . f` will be. (It has to do with detailed semantics of function arity; `undefined` is bottom but `Identity . undefined` is not.)

You may need to use some type annotation on the coercion; I haven't checked.